source/core/output_buffers/ut_output_table_buffer.tpb

```diff
@@ -87,22 +89,24 @@ create or replace type body ut_output_table_buffer is
     lc_long_sleep_time   constant number(1) := 1;     --sleep for 1 s when waiting long
     lc_long_wait_time    constant number(1) := 1;     --waiting more than 1 sec
     l_sleep_time         number(2,1) := lc_short_sleep_time;
-    function get_data_from_buffer return ut_varchar2_rows is
-      l_results        ut_varchar2_rows;
+
+    procedure remove_read_data(a_message_ids ut_integer_list) is
       pragma autonomous_transaction;
     begin
-      delete from (
-        select *
-          from ut_output_buffer_tmp where output_id = self.output_id order by message_id
-        )
-      returning text bulk collect into l_results;
+      delete from ut_output_buffer_tmp a
+       where a.output_id = self.output_id
+         and a.message_id in (select column_value from table(a_message_ids));
       commit;
-      return l_results;
     end;
 
   begin
     loop
-      l_buffer_data := get_data_from_buffer();
+      select a.message_id, ut_output_data_row(a.text, a.item_type)
+        bulk collect into l_message_ids, l_buffer_data
+        from ut_output_buffer_tmp a
+       where a.output_id = self.output_id
+       order by a.message_id;
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Add ASC in order to make the order explicit. ([plsql:OrderByExplicitAscCheck](https://sonarcloud.io/organizations/utplsql/rules#rule_key=plsql%3AOrderByExplicitAscCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=utPLSQL&pullRequest=809&issues=AWgA5Io6_MPaBDn0kbL2&open=AWgA5Io6_MPaBDn0kbL2)

[ISSUE_KEY]: AWgA5Io6_MPaBDn0kbL2